merged button protection features from dev_ux

Joseph-Garvey opened the pull request:
The UI is now a bit tidier and prevents the user from clicking a button they shouldn't be able to.
Should help students not get stuck or messing up the experiment too far in.
This is basically an expansion of the disable buttons feature to check whether the user has loaded an image in addition to having a particle in the table, and that the correct particle is selected.

palvarezc commented:
For our meeting earlier, a possible desirable add on:
- [ ]  launch a "Are you sure?" window when the data later is removed.

Joseph-Garvey commented:
> For our meeting earlier, a possible desirable add on:
> 
> * [ ]   launch a "Are you sure?" window when the data later is removed.

Not sure this will be possible. 
See https://github.com/napari/napari/issues/3466
https://github.com/napari/napari/issues/1570

Although I can't see if anyone has ever tried to hook into (if it's possible to hook into) whatever event is called when napari's delete layer button is called and override it.

Joseph-Garvey commented:
Add "test mode" to bypass  button protection + a test for the button protection.

Joseph-Garvey commented:
- Removed user ability to delete layers by removing the button! (This will break in Napari 0.6.0, reflected in new requirements)
- Removed user ability to add new layers by removing the button! (As above)
- Added logo and copyright.
- Added horizontal layout.
- Implemented widget refactor
- Fixed tests broken by new 4D loading code and the refactor.

Joseph-Garvey commented:
![image](https://github.com/user-attachments/assets/5f5f1bca-3e53-429e-873a-12115ea44813)
![image](https://github.com/user-attachments/assets/8f00c18e-7f6b-4159-a977-ec9b08d7e744)

Joseph-Garvey commented:
Struggling to fix this last test failure. Will take a fresh look tmrw morning but feel free to have a go.

palvarezc commented on src/cavendish_particle_tracks/_widget.py:
```diff
@@ -42,66 +46,90 @@
+"""
+        )
+        print(self.intro_text.text())
+        self.intro_text.setFont(QFont("Consolas", 5))
```
On MacOS this looks a bit weird. Maybe related with this warning
```
WARNING: Populating font family aliases took 158 ms. Replace uses of missing font family "Consolas" with one that exists to avoid this cost.
```
![Screenshot 2024-09-10 at 15 02 22](https://github.com/user-attachments/assets/e6d108ae-6778-4b9c-8390-7f0ba69fc7fc)

palvarezc replied:
No rush to fix this.

palvarezc reviewed (changes requested):
Looks good. Given the user cannot add any layers now, we need to make sure that:
- [x]  `Points` layer is added when the data is loaded.

palvarezc commented:
I encounter a problem that I do not really know how to test for:

- Open the plugin with `python launch_debug.py`
- Load data though the `Load data` dialog
- Create a $\Lambda \to p \pi^-$ particle
- Click `Calculate Decay angles`

And I get a segfault :(

Cannot reproduce when opening the plugin from within napari.

palvarezc commented:
> I encounter a problem that I do not really know how to test for:
> ...
> And I get a segfault :(

It seems this happens in `main` too, so not introduced in this PR. Merging...

palvarezc commented on src/cavendish_particle_tracks/_widget.py:
```diff
@@ -41,66 +46,98 @@
-        self.layout().addWidget(self.rad)
-        self.layout().addWidget(self.lgth)
-        self.layout().addWidget(self.ang)
+        self.viewer.window._qt_viewer.layerButtons.hide()  # This will break in napari 0.6.0
```
```suggestion
        # self.viewer.window._qt_viewer.layerButtons.hide()  # This will break in napari 0.6.0
```

palvarezc replied:
Suggest to comment this out while we still allow for several `Stereoshift` and `Decay angles` layers to be created by repeatedly hitting the corresponding button.

Joseph-Garvey replied:
I've instead fixed this by not allowing the button by not allowing the user to open several stereoshift, mag and decay angle dialogs. It's not a long fix and might as well go for the long-term option if it's only 9 ish lines.

The user can still duplicate the layers of the stereoshift dialog, but this is fixed in the stereoshift PR and so is not corrected here.

Joseph-Garvey commented:
Still can't trace why that patch caused the tests to break. Even undoing the changes hasn't fixed it.
Will check it out with a fresh pair of eyes tomorrow morning.

Joseph-Garvey commented:
See issue #120. Will get user-testing essentials done first.

Joseph-Garvey commented:
@samcunliffe 
test_version_fallback is failing on my pc with msg:
c:\Users\josep\Desktop\Particle Tracks\cavendish-particle-tracks\tests\test_misc.py::test_version_fallback failed: mocker = <pytest_mock.plugin.MockerFixture object at 0x00000195972902E0>

    def test_version_fallback(mocker):
        """Test the versioning fallback (in case setuptools_scm didn't work)"""
        import cavendish_particle_tracks  # fmt: skip
    
>       assert cavendish_particle_tracks.__version__ != "unknown"  # type: ignore[attr-defined]
E       AssertionError: assert 'unknown' != 'unknown'
E        +  where 'unknown' = <module 'cavendish_particle_tracks' from 'C:\\Users\\josep\\Desktop\\Particle Tracks\\cavendish-particle-tracks\\src\\cavendish_particle_tracks\\__init__.py'>.__version__

tests\test_misc.py:11: AssertionError

samcunliffe commented on launch_debug.py:
```diff
@@ -6,6 +6,8 @@
     "cavendish-particle-tracks",
 )
+
+viewer.window.add_dock_widget(dock_widget, area="bottom")
```
Oh nice! Can we make a launcher for the installations in the lab, maybe?

It'd be even nicer if we could have this as a property of the plugin somehow... Like whenever I start `napari` I open the plugin and it's there. I will dig in the plugin docs.

palvarezc commented:
Sorry, moved everything over to #131 and dealt with the conflicts with merge there. I think we can safely close this one.
